### PR TITLE
MINOR: Improve Trogdor external command worker docs

### PR DESCRIPTION
**Pull request by stanislavkozlovski**
I was exploring the code for the new external command worker and saw that it wasn't documented.

This change also contains some very minor code style improvements

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java [hunk 1]
```diff
@@ -389,7 +389,7 @@ public void stop(Platform platform) throws Exception {
             spec.shutdownGracePeriodMs().get() : DEFAULT_SHUTDOWN_GRACE_PERIOD_MS;
         if (!executor.awaitTermination(shutdownGracePeriodMs, TimeUnit.MILLISECONDS)) {
             terminatorActionQueue.add(TerminatorAction.DESTROY_FORCIBLY);
-            executor.awaitTermination(1000, TimeUnit.DAYS);
+            executor.awaitTermination(1, TimeUnit.DAYS);
```
Comment (stanislavkozlovski) [inline tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java:392]: 1000 days is overkill, 1 day is consistent with the other workers

File: tests/bin/external_trogdor_command_example.py [hunk 2]
```diff
@@ -20,19 +20,22 @@
...
 #
 
 if __name__ == '__main__':
     # Read the ExternalCommandWorker start message.
     line = sys.stdin.readline()
     start_message = json.loads(line)
     workload = start_message["workload"]
-    print("Starting external_trogdor_command_example with task id %s, workload %s" \
-        % (start_message["id"], workload))
+    print("Starting external_trogdor_command_example with task id %s, workload %s"
```
Comment (stanislavkozlovski) [inline tests/bin/external_trogdor_command_example.py:31]: the "\" was superfluous here

File: tests/bin/external_trogdor_command_example.py [hunk 3]
```diff
@@ -20,19 +20,22 @@
...
+          % (start_message["id"], workload))
     sys.stdout.flush()
-   `print(json.dumps({"status": "running"}))`
+
+    # pretend to start some workload
+    print(json.dumps({"status": "running"}))
```
Comment (stanislavkozlovski) [inline tests/bin/external_trogdor_command_example.py:36]: the backticks in python 2 are equivalent to calling the `repr()` [function ](https://www.tutorialspoint.com/What-does-the-repr-function-do-in-Python-Object-Oriented-Programming) which returns a string out of the result of the `print()` call. I'm nt sure why this was added initially as it has no effect whatsoever

File: TROGDOR.md [hunk 4]
```diff
@@ -123,6 +123,14 @@ ProcessStopFault stops a process by sending it a SIGSTOP signal.  When the fault
 ### NetworkPartitionFault
 NetworkPartitionFault sets up an artificial network partition between one or more sets of nodes.  Currently, this is implemented using iptables.  The iptables rules are set up on the outbound traffic from the affected nodes.  Therefore, the affected nodes should still be reachable from outside the cluster.
 
+External Processes
+========================================
+Trogdor supports running external processes. This is a generic way to run any process in the system via a configurable command - be it a Python program, bash script, docker image, etc.
```
Comment (yangxi) [inline TROGDOR.md]: Maybe "running arbitrary commands in external processes. This is a generic way to run any configurable command in the Trogdor platform - be it...."?

File: TROGDOR.md [hunk 5]
```diff
@@ -123,6 +123,14 @@ ProcessStopFault stops a process by sending it a SIGSTOP signal.  When the fault
...
+
+### ExternalCommandWorker
+ExternalCommandWorker starts an external command defined by the ExternalCommandSpec. It essentially allows you to run any command on any Trogdor agent node and give it any arbitrary JSON input.
```
Comment (yangxi) [inline TROGDOR.md]: Maybe "ExternalCommandWorker starts an process to execute an external command....", also remove "and give it any arbitary JSON input".

Then in next paragraph where it discusses the communication:

+ The worker communicates with the external process via its stdin, stdout and stderr in a JSON  protocol.

File: tests/bin/external_trogdor_command_example.py [hunk 6]
```diff
@@ -20,19 +20,22 @@
 
 #
 # This is an example of an external script which can be run through Trogdor's
-# ExternalCommandWorker.
+# ExternalCommandWorker. It parses the custom workload JSON and sleeps for the given amount of time.
```
Comment (yangxi) [inline tests/bin/external_trogdor_command_example.py]: Maybe "It sleeps for the given amount of time expressed by the delayMS field in the ExternalCommandSpec.

**Review by yangxi — APPROVED** [review]

**Comment by stanislavkozlovski** [comment]
@cmccabe could you give this PR a look when you have time, please?

File: tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java (hunk 1 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java:392]: In general if this returns and the executor hasn't terminated, we have leaked some threads.  I agree one day should be enough to shut down, but if we actually hit that timeout, there is another bug we have to fix.

File: tests/bin/external_trogdor_command_example.py (hunk 3 above)
Comment (cmccabe) [reply tests/bin/external_trogdor_command_example.py:36]: Heh, I'm not sure how that got in there.  Thanks for removing it.

**Comment by cmccabe** [comment]
@stanislavkozlovski , thanks for taking the initiative to add this to the TROGDOR.md readme!

I think perhaps we need to document the actual protocol between the external process and Trogdor somewhere.  Just saying that it uses "a JSON protocol" doesn't give people enough information to write their own jobs.  We could start by copying the stuff from the JavaDoc for ExternalCommandWorker.java.

**Comment by stanislavkozlovski** [comment]
@cmccabe does this look better now?

File: TROGDOR.md [hunk 7]
```diff
@@ -123,6 +123,18 @@ ProcessStopFault stops a process by sending it a SIGSTOP signal.  When the fault
...
+Trogdor supports running arbitrary commands in external processes. This is a generic way to run any configurable command in the Trogdor framework - be it a Python program, bash script, docker image, etc.
+
+### ExternalCommandWorker
+ExternalCommandWorker starts an external command defined by the ExternalCommandSpec. It essentially allows you to run any command on any Trogdor agent node.
+The worker communicates with the external process via its stdin, stdout and stderr in a JSON protocol.
+The JSON can contain the following fields:
+- status: If the object contains this field, the status of the worker will be set to the given value.
+- error: If the object contains this field, the error of the worker will be set to the given value. Once an error occurs, the external process will be terminated.
+- log: If the object contains this field, a log message will be issued with this text.
+
```
Comment (cmccabe) [inline TROGDOR.md:146]: Can we give an example status message and log message here?  Also, we should spell out that the JSON is communicated over stdout, not stderr..

Finally, we should specify what is passed to the ExternalCommandWorker over stdin on startup.

Comment (stanislavkozlovski) [reply TROGDOR.md:146]: I added some more information. Sorry my imagination isn't really running wild

**Comment by cmccabe** [comment]
LGTM

**Pull request merged**